### PR TITLE
Check store id on edges

### PR DESCRIPTION
**Pull request by davidegrohmann**

### CODE REVIEW

File: community/common/src/main/java/org/neo4j/function/Predicates.java [hunk 1]
```diff
@@ -155,6 +155,23 @@ public static void await( Supplier<Boolean> condition, long timeout, TimeUnit un
         await( condition, timeout, unit, defaultPollInterval, TimeUnit.MILLISECONDS );
     }
 
+    public static void awaitEx( ThrowingSupplier<Boolean, Exception> condition, long timeout, TimeUnit unit )
```
Comment (lutovich) [inline community/common/src/main/java/org/neo4j/function/Predicates.java]: `Ex` -> `Exception`?

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java [hunk 2]
```diff
@@ -58,4 +60,19 @@ public long copyStoreFiles( CoreMember from, StoreFileStreams storeFileStreams )
...
     }
+
+    public StoreId fetchStoreId( CoreMember from ) throws StoreIdDownloadFailedException
+    {
+        CompletableFuture<StoreId> storeIdCompletableFuture = new CompletableFuture<>();
+        coreClient.setStoreIdConsumer( storeIdCompletableFuture::complete );
+        coreClient.requestStoreId( from );
+        try
+        {
+            return storeIdCompletableFuture.get();
```
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java:71]: Is it ok to have an unbounded wait here?

Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java:71]: yes

File: community/common/src/main/java/org/neo4j/function/Predicates.java (hunk 1 above)
Comment (lutovich) [inline community/common/src/main/java/org/neo4j/function/Predicates.java]: This method should be called `await` because it is an overload of the following one.

File: community/common/src/main/java/org/neo4j/function/Predicates.java [hunk 3]
```diff
@@ -155,6 +155,23 @@ public static void await( Supplier<Boolean> condition, long timeout, TimeUnit un
...
+            throws TimeoutException, InterruptedException
+    {
+        int defaultPollInterval = 20;
```
Comment (lutovich) [inline community/common/src/main/java/org/neo4j/function/Predicates.java]: This should be a class level constant and not a local variable.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/CoreClient.java [hunk 4]
```diff
@@ -44,20 +47,23 @@
...
+public abstract class CoreClient extends LifecycleAdapter implements StoreFileReceiver, StoreIdReceiver,
                                                                      StoreFileStreamingCompleteListener,
-                                                                     TxStreamCompleteListener, TxPullResponseListener, CoreSnapshotListener
+                                                                     TxStreamCompleteListener, TxPullResponseListener,
+                                                                     CoreSnapshotListener
 {
     private final PullRequestMonitor pullRequestMonitor;
     private final SenderService senderService;
     private StoreFileStreams storeFileStreams = null;
+    private Consumer<StoreId> storeIdConsumer = null;
```
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/CoreClient.java]: Null is the default value, no need to write it explicitly.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/CoreClient.java [hunk 5]
```diff
@@ -164,6 +181,12 @@ public void onTxReceived( final TxPullResponse tx )
     }
 
     @Override
+    public void onStoreIdReceived( final StoreId storeId )
```
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/CoreClient.java]: This API looks strange and fragile. If `setStoreIdConsumer` was not called than this method will throw NPE. I think `requestStoreId` should just return a `Future`.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/GetStoreIdResponse.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+ */
+package org.neo4j.coreedge.catchup.storecopy.edge;
+
+import java.util.Objects;
+
+import org.neo4j.coreedge.server.StoreId;
+
+public class GetStoreIdResponse
+{
+    private StoreId storeId;
```
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/GetStoreIdResponse.java]: Field can be final.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java (hunk 2 above)
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java:71]: Are you sure? I do not think it is ok to have unbounded wait anywhere because if some issue occurs, thread should not just hang infinitely.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java [hunk 7]
```diff
@@ -58,4 +60,19 @@ public long copyStoreFiles( CoreMember from, StoreFileStreams storeFileStreams )
...
+        }
+        catch ( InterruptedException | ExecutionException e )
```
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java]: It is probably a good idea to restore interruption status for `InterruptedException`

File: enterprise/core-edge/src/test/java/org/neo4j/coreedge/EdgeServerStartupProcessTest.java [hunk 8]
```diff
@@ -41,27 +45,108 @@
...
+        when( clusterTopology.coreMembers() ).thenReturn( asSet( coreMember ) );
+        when( localDatabase.isEmpty() ).thenReturn( true );
+
+        DataSourceManager dataSourceManager = mock( DataSourceManager.class );
+        Lifecycle txPulling = mock( Lifecycle.class );
+
+        EdgeServerStartupProcess edgeServerStartupProcess = new EdgeServerStartupProcess( storeFetcher, localDatabase,
+                txPulling, dataSourceManager, new AlwaysChooseFirstServer( hazelcastTopology ),
+                new ConstantTimeRetryStrategy( 1, MILLISECONDS ), NullLogProvider.getInstance(),
+                mock( EdgeTopologyService.class ), new Config( Collections.emptyMap() ) );
```
Comment (lutovich) [inline enterprise/core-edge/src/test/java/org/neo4j/coreedge/EdgeServerStartupProcessTest.java]: `Config.empty()`

File: enterprise/core-edge/src/test/java/org/neo4j/coreedge/EdgeServerStartupProcessTest.java [hunk 9]
```diff
@@ -41,27 +45,108 @@
...
+
+        when( clusterTopology.coreMembers() ).thenReturn( asSet( coreMember ) );
+
+        DataSourceManager dataSourceManager = mock( DataSourceManager.class );
+        Lifecycle txPulling = mock( Lifecycle.class );
+
+        EdgeServerStartupProcess edgeServerStartupProcess = new EdgeServerStartupProcess( storeFetcher, localDatabase,
+                txPulling, dataSourceManager, new AlwaysChooseFirstServer( hazelcastTopology ),
+                new ConstantTimeRetryStrategy( 1, MILLISECONDS ), NullLogProvider.getInstance(),
+                mock( EdgeTopologyService.class ), new Config( Collections.emptyMap() ) );
```
Comment (lutovich) [inline enterprise/core-edge/src/test/java/org/neo4j/coreedge/EdgeServerStartupProcessTest.java]: `Config.empty()`

File: enterprise/core-edge/src/test/java/org/neo4j/coreedge/EdgeServerStartupProcessTest.java [hunk 10]
```diff
@@ -70,23 +155,25 @@ public void startShouldReplaceLocalStoreWithStoreFromCoreServerAndStartPolling()
...
+        when( localDatabase.isEmpty() ).thenReturn( false );
 
         DataSourceManager dataSourceManager = mock( DataSourceManager.class );
         Lifecycle txPulling = mock( Lifecycle.class );
 
         EdgeServerStartupProcess edgeServerStartupProcess = new EdgeServerStartupProcess( storeFetcher, localDatabase,
                 txPulling, dataSourceManager, new AlwaysChooseFirstServer( hazelcastTopology ),
                 new ConstantTimeRetryStrategy( 1, MILLISECONDS ), NullLogProvider.getInstance(),
-                mock( EdgeTopologyService.class ), config );
+                mock( EdgeTopologyService.class ), new Config( Collections.emptyMap() ) );
```
Comment (lutovich) [inline enterprise/core-edge/src/test/java/org/neo4j/coreedge/EdgeServerStartupProcessTest.java]: `Config.empty()`

File: community/common/src/main/java/org/neo4j/function/Predicates.java (hunk 1 above)
Comment (davidegrohmann) [inline community/common/src/main/java/org/neo4j/function/Predicates.java]: if you do that you need to explicitly cast lambdas to either `Supplier` or `ThrowingSupplier`, so no let's name them differently

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java (hunk 2 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/StoreCopyClient.java:71]: Yes I am sure since everything in around store copy has unbounded waits, so I am not gonna change that in this PR. I might consider to revise all the unbound waits related to store copy and issue a PR for that.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/CoreClient.java (hunk 5 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/CoreClient.java]: No. This is consistent with the rest of the API in the class. If we consider to change the API it will be in a different PR.
